### DISCUSSION TITLE
Updating the power switch on the rover

### DISCUSSION THREAD

**Discussion by apollokit** [nasa-jpl/open-source-rover#235] [Ideas]
Hi all,
This is to report on my just-completed update of the power switch on my rover, and to open a discussion about maybe baselining a new power switch.

The current baseline power switch situation with the rover is not the greatest, because I continually need to pull the battery out when I need to recharge it. I wanted to put a new switch in that would allow me to flip between charging the battery, and powering the control board.

To that end, this is what I did.

**Here's the current configuration for the power switch**:
![image](https://user-images.githubusercontent.com/9410259/101717168-9ddaf080-3a53-11eb-81cc-b72d7031c6a2.png)

**My updated configuration**:

I used a DPDT switch (On-Off-On, also maybe referred to as DP3T, depends where you look), with three states:
1. Battery is powering the control board
2. No connection
3. Battery is connected to an external charger

Note that in the image below, the battery is connected to the orange box to the bottom left with a 1 and 2.

![image](https://user-images.githubusercontent.com/9410259/101717186-a6cbc200-3a53-11eb-981b-83b3b3fb930f.png)

**[Here are images from my build](https://photos.app.goo.gl/SWYQ59ki2Nn6471X9)**

I'm using [this battery](https://www.amazon.com/gp/product/B07YP73LMX/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1&fpw=alm) and [this charger](https://www.amazon.com/gp/product/B07H899K7Y/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1&fpw=alm), so the parts list below reflects that. It wouldn't be hard to change this for different battery connectors though.

**Here's my parts list (~$30 on Amazon):**
1. [Switch](https://www.amazon.com/gp/product/B085ZD3YM7/ref=ppx_yo_dt_b_asin_title_o05_s00?ie=UTF8&psc=1)
2. [Deans plug connectors](https://www.amazon.com/gp/product/B07WHPD4KD/ref=ppx_yo_dt_b_asin_title_o06_s00?ie=UTF8&psc=1)
3. [JST XH Connectors](https://www.amazon.com/gp/product/B08D77WY6K/ref=ppx_yo_dt_b_asin_title_o06_s00?ie=UTF8&psc=1) (an extra connection needed between the battery and the charger)
4. [Twist on wire connectors](https://www.amazon.com/Gardner-Bender-25-AWC-Assortment-Electrical/dp/B0002YUNRE/ref=sr_1_3?dchild=1&keywords=twist%2Bon%2Bwire%2Bconnectors%2Bgardner%2Bbender&qid=1607571671&sr=8-3&th=1)

**I'd like to get people's thoughts on this. Specific questions I have are:**
1. Should we update the baseline rover build to include an upgraded switch like this?
2. Do people like my chosen DPDT state configuration?
3. My usual use case is that I run the rover for a while, and then I take a break while I charge it. I could also see including a couple batteries in the rover and switching between them - one charging, and one powering the control board. Are there others that do this?
4. I ended up just running the charger connectors through holes in the body structure, and every time I want to charge I have to manually connect to those. Maybe there's a more elegant solution for this, but this is what i came up with for first try. Thoughts?

**Comment by SConaway** [comment]
Not a maintainer, but here are some of my opinions where I have them. This is coming from the leader of a high school robotics club which was **trying** to build OSR last school year (we were cost-limited, for example, and even $30 could be hard to find).


1. I would say it would be good to include as an optional task. I personally would have, but not everyone would. (I hate removing batteries to charge things.)
---

4. I think there should be some panel-mount connectors for the charger, no? Also, it could be a panel-mounted AC plug (sealed, not bare wiring) connected to the charger (which lives inside the Rover), possibly. Interested to hear everyone's thoughts on that. Remember, easier is better if you're targeting this to non-EE's.

**Reply by ericjunkins** [reply]
1. Yeah kit, I think it's a good idea to baseline in, it's a very nice convenient thing.

2. I'm still confused about the configuration of the switch. According to your diagram it is DPDT, but earlier you said it explicitly has an off state as well, meaning it's DP3T. Either way, I think DP3T is what we want, so there is an off state, and you aren't accidentally putting voltage to the charging connector when you just want it off.

3. This could start to get complicated if youre trying to have both hooked up at the same time and switch between them. It's not impossible, but I think it's a pretty custom solution we might not get enough return on for doing.

4. Panel mount is the way to go I think. This is what I have on mine to do this btw https://www.amazon.com/Housing-Anderson-Powerpole-Connectors-Resistant/dp/B077VWXXYG/ref=mp_s_a_1_9?dchild=1&keywords=Anderson+powerpole&qid=16075796

I'm hesitant to recommend anyone do anything directly with AC power, I think it's better to stick with things that are going to be a little bit safer

**Reply by dcschooley** [reply]
Here's what I use to mount the PowerPoles. The small tab in the center fits into the slots on the PowerPoles and holds them in place.  You need two of these. The holder gets a 4-40 heat-set insert and you screw it in from the front. 

![Screen Shot 2020-12-10 at 1 32 05 PM](https://user-images.githubusercontent.com/59415821/101820434-3a50d180-3aec-11eb-8442-362011955125.png)

**Reply by apollokit** [reply]
1. So that's 1 vote for baselining, 1 vote against
2. @ericjunkins So [this source](https://www.littelfuse.com/technical-resources/technical-centers/commercial-vehicle-technical-center/poles-and-throws.aspx) is what i'm basing my DPDT terminology on. They call this switch a DPDT On-Off-On. 
3. Agreed that baselining two batteries and switching between them probably doesn't offer enough return
4. yeah, agreed panel mount is better. My current method is super jank. 
I like Eric's [link](https://www.amazon.com/Housing-Anderson-Powerpole-Connectors-Resistant/dp/B077VWXXYG/ref=mp_s_a_1_9?dchild=1&keywords=Anderson+powerpole&qid=16075796), though we probably only want one set of power pole connectors, right? And yeah, AC seems scary.
@dcschooley , any chance you could share a picture of your mounting for the power poles? I'm not really groking it from that CAD drawing...

**Reply by dcschooley** [reply]
I've include a picture of the mounted PowePoles and another of a PowerPole by itself. My 3D printed pieces grab the PowerPoles from the top and bottom. The PowerPole picture shows the slot that matches up with the tab on the mounting piece. 

![PowerPole Retention Annotated](https://user-images.githubusercontent.com/59415821/101859735-e9fc6280-3b31-11eb-9f23-21b8b3e13389.png)
 

![tempImageHZFmjF](https://user-images.githubusercontent.com/59415821/101858861-1616e400-3b30-11eb-921d-513573fa207b.gif)
![tempImagekHNtNW](https://user-images.githubusercontent.com/59415821/101858871-1ca55b80-3b30-11eb-82cd-c6b3f4149ec4.gif)

**Reply by dcschooley** [reply]
I'm for baselining this if we can keep the cost down. People are attempting to solve this problem independently, and nobody likes opening up the body to charge the battery. PowerPoles are great but they might not be a good default option. The parts themselves are fairly cheap, but you need a good crimper. The PowerWerx crimper that makes perfect PowerPole crimps is $45. That's a lot to install perhaps six PowerPoles. You could also go with a high-quality generic crimper, but you won't save much. You want your connections to be foolproof when working with LIPOs. 

PowerWerx does sell some PowerPoles that already have leads attached, and batteriesamerica sells PowerPole to Tamiya adapters. You could buy a few of those and patch something together, but the price would accumulate fairly quickly. 

A 3D-printed panel mount for the Tamiya connectors might be possible. A baseline, panel-mounted Tamiya connector might be best because Tamiya connectors are already part of the baseline. A DPDT switch isn't much more expensive than a SPST switch.

**Comment by dcschooley** [comment]
You have to be careful with the RoboClaws because they expect the battery to remain in the circuit when the switch is turned off. The RoboClaws use regenerative braking to dump excess energy from the motors back into the battery. If the battery is disconnected when the motors are running, then the energy doesn't have anywhere to go and the RoboClaws can be damaged. That's the reason for the diode+fuse combination and the reason I went with two switches. Flipping the switch when the rover is cruising across the prairie isn't likely, but it's something that could happen a lot during testing. I use a DPDT on-off-on switch to move the battery back and forth between the charger connection and the board. The 'off' part of the switch isn't necessary but I kind of like it. The second switch operates the same way as the single switch in the baseline design.

**Reply by ericjunkins** [reply]
The "off" position in this case can connect the roboclaws to GND through a big bleed resistor, which would be the path to dissipate any regenerative energy if necessary

**Reply by dcschooley** [reply]
Let me think about that a bit. Another possibility might be to tie all of the grounds together. (The lower three dots in Kit's drawing.) I'm a bit hesitant to deviate too far from the RoboClaw documentation, and keeping the grounds tied together is consistent with the admonition to always disconnect the positive battery terminal first. I think that will work and allow you to do it without the second switch.

**Reply by apollokit** [reply]
Interesting. I hadn't thought about regenerative braking, good point. I'm not personally so worried about flipping the switch and disconnecting the battery while the motors are turning (which should be the only time they're able to dump energy back into the controllers, correct?), but that's probably not a safe baseline for everybody.

**Reply by dcschooley** [reply]
Flipping the switch while the motors are running would be quite rare. I think it would almost always happen during testing when the motors start going fast enough that someone gets excited and throws the switch. If the excitement is caused by corner motors getting jammed into the stops, then there is not enough stored energy to worry about. 

I am back to thinking there needs to be a second switch per the diagram on page 41 of the RoboClaw manual. Your connection bypasses that 10k resistor (R1 in the RoboClaw docs) completely. R1 does some important things.

**Reply by ericjunkins** [reply]
I don't think two switches are necessary to accomplish this

**Reply by dcschooley** [reply]
> The "off" position in this case can connect the roboclaws to GND through a big bleed resistor, which would be the path to dissipate any regenerative energy if necessary

There's no connection at all in the off position of the DPDT on-off-on switches. The off position is just a holding spot between the two positions where the switch actually makes connections. One of the reasons I like the DPDT on-off-on switches is that you can completely isolate the battery from everything. 

We basically have three states: (I'll call the switch we have now the "board switch" for the lack of anything better.)
1) Board switch is closed with battery connected to board. (Rover can drive.)
2) Board switch is open with battery connected to board. (Rover is off and all of the RoboClaw protection works. You need to be in this state for a short period of time before closing the board switch to drive and after opening the board switch.)
3) Board switch is open with battery connected to charger. 

Our difficulty is that the battery currently doesn't share any common points with the board switch. Maybe we swap the locations of the switch and the diode/fuse combination in the circuit so that the switch connects to the positive terminal of the battery. I think you could probably do it with a 4-pole, triple-throw switch. I've attached a drawing that might work.

![Maybe this works](https://user-images.githubusercontent.com/59415821/101956783-caaf1500-3bc5-11eb-86d6-d2dfd88afd40.png)

**Reply by dcschooley** [reply]
Another option might be to use a true DP3T switch and swap the positions of the board switch/R1 combination and the diode/fuse combination in the circuit. You then use the switch to move the positive terminal of the battery to one terminal or the other of R1. This would require a permanent connection between the board and the negative terminal of the battery. It also doesn't completely meet state #2, but you might be able to do that with and additional resistor. I'm not a fan of this option.
